src/lib/mongoBackend/mongoRegistrationGet.cpp

```diff
@@ -0,0 +1,325 @@
+*
+* setRegistrationId -
+*/
+static void setRegistrationId(ngsiv2::Registration* regP, const mongo::BSONObj& r)
+{
+  regP->id = getFieldF(r, "_id").OID().toString();
+}
+
+
+
+/* ****************************************************************************
+*
+* setDescription -
+*/
+static void setDescription(ngsiv2::Registration* regP, const mongo::BSONObj& r)
+{
+  if (r.hasField(REG_DESCRIPTION))
+  {
+    regP->description         = getStringFieldF(r, REG_DESCRIPTION);
+    regP->descriptionProvided = true;
+  }
+  else
+  {
+    regP->description         = "";
+    regP->descriptionProvided = false;
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* setProvider -
+*/
+static void setProvider(ngsiv2::Registration* regP, const mongo::BSONObj& r)
+{
+  regP->provider.url = (r.hasField(REG_PROVIDING_APPLICATION))? getStringFieldF(r, REG_PROVIDING_APPLICATION): "";
+}
+
+
+
+/* ****************************************************************************
+*
+* setDataProvided -
+*
+* Make sure there is only ONE "contextRegistration" in the vector
+* If we have more than one, then the Registration is made in API V1 as this is not
+* possible in V2 and we cannot respond to the request using the current implementation of V2
+*
+*/
+static bool setDataProvided(ngsiv2::Registration* regP, const mongo::BSONObj& r, bool arrayAllowed)
+{
+  std::vector<mongo::BSONElement> crV = getFieldF(r, REG_CONTEXT_REGISTRATION).Array();
+
+  if (crV.size() > 1)
+  {
+    return false;
+  }
+
+  //
+  // Extract first CR from the contextRegistration vector
+  //
+  mongo::BSONObj cr0 = crV[0].embeddedObject();
+
+
+
+  //
+  // Entities
+  //
+  std::vector<mongo::BSONElement>  dbEntityV = getFieldF(cr0, REG_ENTITIES).Array();
+
+  for (unsigned int ix = 0; ix < dbEntityV.size(); ++ix)
+  {
+    ngsiv2::EntID    entity;
+    mongo::BSONObj   ce = dbEntityV[ix].embeddedObject();
+
+    if (ce.hasField(REG_ENTITY_ISPATTERN))
+    {
+      std::string isPattern = getStringFieldF(ce, REG_ENTITY_ISPATTERN);
+
+      if (isPattern == "true")
+      {
+        entity.idPattern = getStringFieldF(ce, REG_ENTITY_ID);
+      }
+      else
+      {
+        entity.id = getStringFieldF(ce, REG_ENTITY_ID);
+      }
+    }
+    else
+    {
+      entity.id = getStringFieldF(ce, REG_ENTITY_ID);
+    }
+    
+    if (ce.hasField(REG_ENTITY_ISTYPEPATTERN))
+    {
+      std::string isPattern = getStringFieldF(ce, REG_ENTITY_ISTYPEPATTERN);
+
+      if (isPattern == "true")
+      {
+        entity.typePattern = getStringFieldF(ce, REG_ENTITY_TYPE);
+      }
+      else
+      {
+        entity.type = getStringFieldF(ce, REG_ENTITY_TYPE);
+      }
+    }
+    else
+    {
+      entity.type = getStringFieldF(ce, REG_ENTITY_TYPE);
+    }
+
+    regP->dataProvided.entities.push_back(entity);
+  }
+
+
+  //
+  // Attributes
+  //
+  std::vector<mongo::BSONElement> dbAttributeV = getFieldF(cr0, REG_ATTRS).Array();
```

Comment:
Maybe the logic for "filling" attributes could be done in a separated function, to make setDataProvider() shorter and easier.

Reply:
Definitely

Reply:
Fixed in f114c3e